### PR TITLE
Expedite delayed operations when tests are waiting

### PR DESCRIPTION
**Pull request by sharwell**
Saves about 20 minutes off our longest test legs.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+using System.Diagnostics;
+using System.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Utilities
+{
+    /// <summary>
+    /// Extensions to <see cref="CancellationToken"/>.
+    /// </summary>
+    internal static class CancellationTokenExtensions
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: 📝 Taken from Microsoft/vs-threading. This class avoids creation of a linked token source when one of the inputs is not cancellable. In Roslyn production runs, this is _always_ the case, so the behavior provided by `CombineWith` is required for preventing some amount of performance regression in the asynchronous processing queues.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs [hunk 2]
```diff
@@ -17,6 +17,8 @@ internal sealed partial class AsynchronousOperationListener : IAsynchronousOpera
 
         private readonly string _featureName;
         private readonly HashSet<TaskCompletionSource<bool>> _pendingTasks = new HashSet<TaskCompletionSource<bool>>();
+        private CancellationTokenSource _blockedOnCompletionTokenSource = new CancellationTokenSource();
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: either here or in constructor? rather than both?

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs [hunk 3]
```diff
@@ -94,13 +102,24 @@ private void Decrement_NoLock(AsyncToken token)
             }
         }
 
-        public Task CreateWaitTask()
+        public Task CreateWaitTask(bool willBlockOnCompletion)
         {
+            if (willBlockOnCompletion && _counter > 0)
+            {
+                _blockedOnCompletionTokenSource.Cancel();
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: I don't understand this? so, creating new wait task cancels all waiting. so, will bock on completion make it not block on delay? who use "will bock on completion : false" ?

Comment (heejaechang) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: so, this is only case where it get cancelled?

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs [hunk 4]
```diff
@@ -94,13 +102,24 @@ private void Decrement_NoLock(AsyncToken token)
...
+            }
+
             using (_gate.DisposableWait(CancellationToken.None))
             {
                 if (_counter == 0)
                 {
                     // There is nothing to wait for, so we are immediately done
+                    if (_blockedOnCompletionTokenSource.IsCancellationRequested)
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: I am not sure I understand this. so.. this is the case where CreateWaitTask is called while _count > 0 and then later CreateWaitTask is called again once _count == 0 .. and then for this to work as epxected, it is expected someone to call CreateWaitTask again?

why, does this require to reset cancellation? once cancellation is called, shouldn't thing to stay as cancelled?

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 5]
```diff
@@ -65,9 +66,23 @@ protected async Task WaitForIdleAsync()
...
+                    {
+                        // TODO: will safestart/unwarp capture cancellation exception?
+                        var timeLeft = BackOffTimeSpanInMS - diffInMS;
+                        await Task.Delay(Math.Max(MinimumDelayInMS, timeLeft), combinedCancellationToken.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException) when (blockedOnCompletion.IsCancellationRequested && !CancellationToken.IsCancellationRequested)
+                    {
+                        // The delay terminated early to accommodate a blocking operation. Make sure to delay long
+                        // enough that low priority (on idle) operations get a chance to be triggered.
+                        await Task.Delay(TimeSpan.FromMilliseconds(10)).ConfigureAwait(false);
```
Comment (heejaechang) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: if we don't have this waht happen?

Comment (heejaechang) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: I would also add that if blockedOnCompletion.IsCancellationToken.IsCancelled, then return right away rather than do any of waiting.

Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: > if we don't have this waht happen?

The containing loop spins until WaitForAllOperationsAsync times out. I'm not sure what was getting blocked.

File: src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs [hunk 6]
```diff
@@ -62,7 +62,7 @@ internal async Task<IList<ITagSpan<IErrorTag>>> GetErrorsFromUpdateSource(TestWo
                 {
                     source.RaiseDiagnosticsUpdated(updateArgs);
 
-                    await wrapper.WaitForTags();
+                    await wrapper.WaitForTags(willBlockOnCompletion: true);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs]: everything i've seen so far seems to be passing in 'true' here.  Could you give an intuition on who should *not* pass in true?

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        {
+            if (original.IsCancellationRequested || !other.CanBeCanceled)
+            {
+                return new CombinedCancellationToken(original);
+            }
+
+            if (other.IsCancellationRequested || !original.CanBeCanceled)
+            {
+                return new CombinedCancellationToken(other);
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: dumb question.  why do you need to check the .CanBeCanceled part?  if you already know that one or the other has had cancellation requested, why do you ever need to check the other.

The docs say "Creates a new <see cref="CancellationToken"/> that is canceled when **any** of a set of other tokens are canceled."  Since you already know "**any**" token was canceled, why are the others relevant at all?

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        /// <param name="others">The additional tokens.</param>
+        /// <returns>A struct that contains the combined <see cref="CancellationToken"/> and a means to release memory when you're done using it.</returns>
+        public static CombinedCancellationToken CombineWith(this CancellationToken original, params CancellationToken[] others)
+        {
+            Contract.ThrowIfNull(others, nameof(others));
+
+            if (original.IsCancellationRequested)
+            {
+                return new CombinedCancellationToken(original);
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: this seems to have different semantics from above.

File: src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticTaggerWrapper.cs [hunk 9]
```diff
@@ -128,15 +128,15 @@ public void Dispose()
             _registrationService.Unregister(_workspace);
         }
 
-        public async Task WaitForTags()
+        public async Task WaitForTags(bool willBlockOnCompletion)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticTaggerWrapper.cs]: i can't find anyone who passes false in for this.

File: src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs]: If for some reason you want to wait on items without changing runtime behavior. The thing I didn't like is something that looks like a simple observer producing behavior changes.

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 7 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: If `original` cannot be cancelled, then `CancellationTokenSource.CreateLinkedSource(other, original).Token` is observably equivalent to `other.Token`, so we avoid the allocation. The same is true when `other` is already cancelled. Based on the wording of your question, it sounds like you may have misread the condition as `&&` when it's `||`. Can you see if this answers your questions?

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 8 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: Can you describe an input pair (CancellationToken a, CancellationToken b) for which the following are not the same?

* `CombineWith(a, b)`
* `CombineWith(a, new[] { b })`

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: Indeed!  I did misread it in that exact manner.  Thanks for the clarification!

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: Retracted.  Misread things

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: Retracted.  Misread things

File: src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs]: > If for some reason you want to wait on items without changing runtime behavior. The thing I didn't like is something that looks like a simple observer producing behavior changes. I wasn't especially happy with the design choice so alternative solutions would be welcome.

I think i'd prefer to just change the name here to something that makes that more clear.  i.e. roll this into the name of hte method.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: Fixed

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: > so, creating new wait task cancels all waiting. so, will bock on completion make it not block on delay?

I don't understand the question. Can you rephrase it?

> who use "will bock on completion : false" ?

Removed this parameter

> so, this is only case where it get cancelled?

Yes

File: src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/TestUtilities/Squiggles/AbstractSquiggleProducerTests.cs]: Renamed the main wait method and removed the `willBlockOnCompletion` parameters.

File: src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticTaggerWrapper.cs (hunk 9 above)
Comment (sharwell) [reply src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticTaggerWrapper.cs]: Parameter removed.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: This handles a race condition:

1. Operations exist, so the token source is marked as blocked
2. Operations complete
3. The semaphore is acquired
4. We find out that no operations need to run, so we return early

**Comment by sharwell** [comment]
@heejaechang @CyrusNajmabadi this is ready for review again

File: src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationWaiter.cs [hunk 10]
```diff
@@ -10,7 +10,12 @@ internal interface IAsynchronousOperationWaiter
         bool TrackActiveTokens { get; set; }
         ImmutableArray<AsynchronousOperationListener.DiagnosticAsyncToken> ActiveDiagnosticTokens { get; }
 
-        Task CreateWaitTask();
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationWaiter.cs]: We've confirmed this isn't being used in any test harnesses anywhere else that is driving Roslyn? I think some of that is hidden behind the waiting service but we should confirm that.

File: src/Workspaces/Core/Portable/Shared/TestHooks/INotifyBlockedOnCompletion.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis.Shared.TestHooks
+{
+    internal interface INotifyBlockedOnCompletion
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/TestHooks/INotifyBlockedOnCompletion.cs]: Please document. My first take on the name was "completion" was this was also related to the completion feature. In any other project the name might be fine but it's a bit more confusing here. :smile:

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 8 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: FWIW I had the same misread. :smile:

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: Tests? If we're copying code we should be copying tests too.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs [hunk 12]
```diff
@@ -72,6 +78,8 @@ private void Decrement_NoLock(AsyncToken token)
                 }
 
                 _pendingTasks.Clear();
+                _blockedOnCompletionTokenSource = new CancellationTokenSource();
+                _blockedOnCompletionToken = _blockedOnCompletionTokenSource.Token;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: Is there ever a case where _blockedOnCompletionToken isn't just equal to _blockedOnCompletionTokenSource.Token? My assumption is if there are two fields there some subtlety why it's necessary...

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.AsyncToken.cs [hunk 13]
```diff
@@ -20,6 +20,8 @@ public AsyncToken(AsynchronousOperationListener listener)
                 listener.Increment_NoLock();
             }
 
+            public CancellationToken BlockedOnCompletion => _listener?.BlockedOnCompletion ?? default;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.AsyncToken.cs]: Is there a reason this should be conceptually associated with a token or is just for convenience because sometimes we pass around a token and not the originating listener?

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 14]
```diff
@@ -65,9 +66,23 @@ protected async Task WaitForIdleAsync()
...
+                var blockedOnCompletion = notifyBlockedOnCompletion.BlockedOnCompletion;
+                using (var combinedCancellationToken = CancellationToken.CombineWith(blockedOnCompletion))
+                {
+                    try
+                    {
+                        // TODO: will safestart/unwarp capture cancellation exception?
+                        var timeLeft = BackOffTimeSpanInMS - diffInMS;
+                        await Task.Delay(Math.Max(MinimumDelayInMS, timeLeft), combinedCancellationToken.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException) when (blockedOnCompletion.IsCancellationRequested && !CancellationToken.IsCancellationRequested)
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: I'm really wishing the INotifyBlockedOnCompletion interface was instead an interface that gave you a Task.Delay-like method, and implemented the cancellation logic internally. More concretely:

```
interface IExpeditableDelaySource
{
    Task CreateExpeditableDelay(TimeSpan span, CancellationToken cancellationToken);
}
```

Then it means that the catching of the `BlockedOnCompletion` OperationCancelledException can be abstracted inside of it (it just is treated as a Delay that just completed sooner), which would greatly simplify this logic and also the logic at the other uses. It also means any new use doesn't have a chance of accidentally forgetting to dispose the combined token, won't incorrectly catch the OperationCanceled, etc.

Comment (jasonmalinowski) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: (and if you need to distinguish which was the cause of the delay, have it return a bool, which can be ignored.)

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs (hunk 5 above)
Comment (jasonmalinowski) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: Can you expand on this comment with the additional comments above?

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 15]
```diff
@@ -85,10 +100,10 @@ private async Task ProcessAsync()
                     // wait for next item available
                     await WaitAsync(this.CancellationToken).ConfigureAwait(continueOnCapturedContext: false);
 
-                    using (this.Listener.BeginAsyncOperation("ProcessAsync"))
+                    using (var asyncToken = this.Listener.BeginAsyncOperation("ProcessAsync"))
                     {
                         // we have items but workspace is busy. wait for idle.
-                        await WaitForIdleAsync().ConfigureAwait(continueOnCapturedContext: false);
+                        await WaitForIdleAsync(asyncToken).ConfigureAwait(continueOnCapturedContext: false);
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: Mentally it's very strange to me we're passing the token here and not the listener. It implies that somehow to me the wait is dependent on the token which isn't the case at all.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 4 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: @sharwell Is that special case avoided if we just put the cancelling of the token inside the gate? That seems easier and given it's not like the code a few lines earlier is avoiding taking the lock we're reducing any contention in practice. It also means you could just put the cancel by the code that's adding the task to _pendingTasks, which then means it's easier to audit this -- all code that adds or removes from _pendingTasks should also be manipulating this token in some way, since you're just maintaining the invariant that the token is cancelled if _pendingTasks is non-empty.

(if you _really_ want to leave it this way, you need a very long comment here.)

File: src/Workspaces/Core/Portable/Shared/TestHooks/INotifyBlockedOnCompletion.cs (hunk 11 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Shared/TestHooks/INotifyBlockedOnCompletion.cs]: Also before you document this, see my earlier comment where I'm describing what I think we actually *want* this interface to be.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs [hunk 16]
```diff
@@ -17,6 +17,8 @@ internal sealed partial class AsynchronousOperationListener : IAsynchronousOpera
 
         private readonly string _featureName;
         private readonly HashSet<TaskCompletionSource<bool>> _pendingTasks = new HashSet<TaskCompletionSource<bool>>();
+        private CancellationTokenSource _blockedOnCompletionTokenSource;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: See other comments first, but please comment this.

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: ➡️ This type has now been removed

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 12 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: ➡️ Now only one field

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs (hunk 14 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: 💯 outstanding

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs (hunk 15 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: ➡️ Now passing the listener as an `IExpeditableDelaySource`

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 16 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: This should be clearer now, since the use of the cancellation token source is contained in same file with the source itself.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: It _was_ special because it needed asynchronous cancellation to avoid throwing in an attempt to re-enter the gate. I moved it inside the gate now and used `CancelAfter` instead of `Cancel` to ensure the continuations are not synchronously invoked.

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs (hunk 5 above)
Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: Expanded to the degree known.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: This should be much clearer now.

File: src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs]: Type has been removed.

File: src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationWaiter.cs (hunk 10 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/IAsynchronousOperationWaiter.cs]: I have refactored this in the past without problems, so I'm assuming that hasn't changed.

File: src/Workspaces/Core/Portable/Shared/TestHooks/INotifyBlockedOnCompletion.cs (hunk 11 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/TestHooks/INotifyBlockedOnCompletion.cs]: Yep, you was right

**Comment by sharwell** [comment]
@jasonmalinowski @heejaechang @CyrusNajmabadi this has been substantially simplified, please take another look :smile:

File: src/EditorFeatures/Core/Shared/Utilities/ResettableDelay.cs [hunk 17]
```diff
@@ -52,8 +55,16 @@ private async Task StartTimerAsync(bool continueOnCapturedContext)
...
             do
             {
-                // Keep delaying until at least delayInMilliseconds has elapsed since lastSetTime 
-                await Task.Delay(_delayInMilliseconds).ConfigureAwait(continueOnCapturedContext);
+                try
+                {
+                    // Keep delaying until at least delayInMilliseconds has elapsed since lastSetTime
+                    await _expeditableDelaySource.Delay(TimeSpan.FromMilliseconds(_delayInMilliseconds), CancellationToken.None).ConfigureAwait(continueOnCapturedContext);
+                }
+                catch (OperationCanceledException)
```
Comment (sharwell) [inline src/EditorFeatures/Core/Shared/Utilities/ResettableDelay.cs]: 🐛 This is preventing callers from figuring out if the task has been expedited.

File: src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs [hunk 18]
```diff
@@ -30,9 +31,26 @@ internal sealed partial class AsynchronousOperationListener : IAsynchronousOpera
...
+        public async Task Delay(TimeSpan delay, CancellationToken cancellationToken)
+        {
+            var expeditedDelayCancellationToken = _expeditedDelayCancellationTokenSource.Token;
+            using (var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, expeditedDelayCancellationToken))
+            {
+                try
+                {
+                    await Task.Delay(delay, cancellationTokenSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException) when (expeditedDelayCancellationToken.IsCancellationRequested && !cancellationToken.IsCancellationRequested)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs]: 🐛 This is preventing callers from figuring out if the task has been expedited.

File: src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs [hunk 19]
```diff
@@ -2343,7 +2343,7 @@ public async Task TestCreateWithBufferNotInWorkspace()
                         }
 
                         var waiter = listenerProvider.GetWaiter(FeatureAttribute.Classification);
-                        await waiter.CreateWaitTask();
+                        await waiter.CreateExpeditedWaitTask();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs]: great name :)

File: src/EditorFeatures/Core/Shared/Utilities/ResettableDelay.cs [hunk 20]
```diff
@@ -52,8 +55,12 @@ private async Task StartTimerAsync(bool continueOnCapturedContext)
         {
             do
             {
-                // Keep delaying until at least delayInMilliseconds has elapsed since lastSetTime 
-                await Task.Delay(_delayInMilliseconds).ConfigureAwait(continueOnCapturedContext);
+                // Keep delaying until at least delayInMilliseconds has elapsed since lastSetTime
+                if (!await _expeditableDelaySource.Delay(TimeSpan.FromMilliseconds(_delayInMilliseconds), CancellationToken.None).ConfigureAwait(continueOnCapturedContext))
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Utilities/ResettableDelay.cs]: such a nifty idea.

File: src/Workspaces/Core/Portable/Shared/TestHooks/IExpeditableDelaySource.cs [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+        /// <param name="delay">The time to wait before completing the returned task, or <c>TimeSpan.FromMilliseconds(-1)</c> to wait indefinitely.</param>
+        /// <param name="cancellationToken">A cancellation token to observe while waiting for the task to complete.</param>
+        /// <returns><see langword="true"/> if the delay compeleted normally; otherwise, <see langword="false"/> if the delay completed due to a request to expedite the delay.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <para><paramref name="delay"/> represents a negative time interval other than <c>TimeSpan.FromMilliseconds(-1)</c>.</para>
+        /// <para>-or-</para>
+        /// <para>The <paramref name="delay"/> argument's <see cref="TimeSpan.TotalMilliseconds"/> property is greater than <see cref="int.MaxValue"/>.</para>
+        /// </exception>
+        /// <exception cref="OperationCanceledException">The delay has been canceled.</exception>
+        Task<bool> Delay(TimeSpan delay, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/TestHooks/IExpeditableDelaySource.cs]: Great docs as always!

**Comment by CyrusNajmabadi** [comment]
> Saves about 20 minutes off our longest test legs.

Wow.  Wow wow wow.  And kudos on achieving this and having a really clean and simple design here.  I'm with Jason in that i'd lke to avoid `fancy threading` as much as possible due to the costs when things invariably go wrong.

File: src/EditorFeatures/Core/Shared/Utilities/ResettableDelay.cs (hunk 20 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Shared/Utilities/ResettableDelay.cs]: I'll take credit for the concept but @jasonmalinowski gets all the credit for making a suggestion that kept it this clean.
